### PR TITLE
Changed fallback logic.

### PR DESCRIPTION
**Pull request by mariacarmina**
## Description

Change the fallback logic when it comes to last block processing.

## Is this PR related with an open issue?

Related to Issue #699 

## Types of changes

- [X] Bug fix (non-breaking change which fixes an issue)
 
## Checklist:

- [X] Follows the code style of this project.

### CODE REVIEW

File: .github/workflows/pytest.yml [hunk 1]
```diff
@@ -43,6 +43,7 @@ jobs:
           stack-version: 7.14.2
       - name: Test with pytest
         run: |
+            bash exportlocal.sh
```
Comment (DMats) [inline .github/workflows/pytest.yml:46]: Interesting.  This line cements the `exportlocal.sh` script as **the** place to specify envvars for the pytest Github check.  And since it's a shell script, can be easily run locally as well. 

Part of me thinks we should specify these variables under the `env` tag so that they're more visible when viewing the pytest.yml file. This would decouple the CI environment from developers local environments.

Comment (calina-c) [reply .github/workflows/pytest.yml:46]: no, please do not do this. Use the pytest.ini file to define ant variables for pytest

File: aquarius/app/cached_block.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+def get_cached_block():
+    """
+    Retrieves the latest block value.
+    :return: `cached_block` stored in the database
+    """
+    return cache.get("cached_block")
+
+
+def update_cached_block(cached_block: Optional[int]):
```
Comment (calina-c) [inline aquarius/app/cached_block.py:32]: or, instead of making it Optional, make it a required int. Then you can remove the None check.

File: aquarius/app/cached_block.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    """
+    Updates the value of `cached_block` in the database
+    :param: cached_block
+    """
+    if cached_block is None:
+        msg = "Cached block is None."
+        logger.error(msg)
+        raise Exception(msg)
+
+    block = get_or_create_cached_block(cached_block)
```
Comment (calina-c) [inline aquarius/app/cached_block.py:42]: you don't need this function. Please look at what you are copying and how you can adapt. The Nonce was "get_or_create" because we needed to initialize it when empty, depending on each user. Here, we can just set the cached block, simply.

File: aquarius/app/cached_block.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    },
+)
+
+
+def get_cached_block():
+    """
+    Retrieves the latest block value.
+    :return: `cached_block` stored in the database
+    """
+    return cache.get("cached_block")
```
Comment (calina-c) [inline aquarius/app/cached_block.py:29]: you can return 0 if it is not set, saving you from an if statement in the main logic

File: aquarius/events/events_monitor.py [hunk 5]
```diff
@@ -265,17 +265,29 @@ def handle_token_uri_update(self, from_block, to_block):
                 )
 
     def get_last_processed_block(self):
-        block = 0
         try:
+            # Initialize the cached block with 0 if cache is empty.
+            if get_cached_block() is None:
```
Comment (calina-c) [inline aquarius/events/events_monitor.py:270]: should the cached block be updated when the ES one is updated? i.e. when doing "store_last_processed_block"?
what happens if the REDIS connection is undefined? Does this code still work as expected (no caching, starts from 0)?

don't you think the existence of REDIS connection for block caching forces the nonce to use REDIS instead of sqlite? What if the user wants to define REDIS just for block caching, but wants to keep sqlite nonce generation? Is that a possibility?

File: aquarius/events/events_monitor.py [hunk 6]
```diff
@@ -265,17 +265,29 @@ def handle_token_uri_update(self, from_block, to_block):
...
+                logging.error("Trying to connect to ES...")
+                time.sleep(5)
             last_block_record = self._es_instance.es.get(
                 index=self._other_db_index, id=self._index_name, doc_type="_doc"
             )["_source"]
-            block = last_block_record["last_block"]
+            # If blocks indexing starts from 0, get the cached last block instead.
+            block = (
+                last_block_record["last_block"]
+                if last_block_record["last_block"] != 0
```
Comment (calina-c) [inline aquarius/events/events_monitor.py:282]: are you sure the value will be an int? Could it possibly be "0" instead?

File: tests/test_events.py [hunk 7]
```diff
@@ -224,16 +227,16 @@ def test_process_block_range(client, base_ddo_url, events_object):
         assert events_object.process_current_blocks() is None
 
 
-def test_get_last_processed_block(events_object):
-    start_block = events_object._start_block
+def test_get_last_processed_block(monkeypatch, events_object):
+    monkeypatch.setenv("REDIS_CONNECTION", "redis://172.15.0.18:6379")
```
Comment (calina-c) [inline tests/test_events.py:231]: you need to test more cases. One of the cases I suggested was the one where ES works properly, then fails only at the n-th call. You can see examples of how to do that in the provider algo validation tests.

File: aquarius/events/events_monitor.py (hunk 6 above)
Comment (mariacarmina) [reply aquarius/events/events_monitor.py]: I have a test for that as well and it works as an int value

File: .github/workflows/pytest.yml (hunk 1 above)
Comment (mariacarmina) [reply .github/workflows/pytest.yml:46]: ok I will delete it

File: setup.py [hunk 8]
```diff
@@ -37,6 +38,7 @@
     "pyshacl==0.18.1",
     "gql==3.0.0",
     "aiohttp==3.8.1",
+    "redis==4.0.2",
```
Comment (calina-c) [inline setup.py:41]: no need for redis and caching anymore, if we are not caching anything

File: tests/test_events.py [hunk 9]
```diff
@@ -224,16 +228,32 @@ def test_process_block_range(client, base_ddo_url, events_object):
...
+        action_thread.join()
+        assert "Connection to ES failed. Trying to connect to back..." in caplog.text
+
+
 def test_get_last_processed_block(events_object):
-    start_block = events_object._start_block
     with patch("elasticsearch.Elasticsearch.get") as mock:
         mock.side_effect = Exception("Boom!")
-        assert events_object.get_last_processed_block() == start_block
+        assert events_object.get_last_processed_block() == 0
```
Comment (calina-c) [inline tests/test_events.py:251]: well, not really. We can have a custom start block which is not 0. The only reason it's 0 for ganache it's because we define it to be. Look into the list of env vars and you'll find what I'm talking about.

**Review by calina-c — CHANGES_REQUESTED** [review]
A few changes still needed.

File: setup.py (hunk 8 above)
Comment (mariacarmina) [reply setup.py:41]: Forgot to delete it

File: tests/test_events.py [hunk 10]
```diff
@@ -224,16 +228,32 @@ def test_process_block_range(client, base_ddo_url, events_object):
...
+            assert events_object.get_last_processed_block() == 24
+
+    with patch("elasticsearch.Elasticsearch.ping") as es_mock:
+        es_mock.return_value = False
+        action_thread = threading.Thread(target=events_object.get_last_processed_block)
+        action_thread.start()
+        time.sleep(5)
+        es_mock.return_value = True
+        action_thread.join()
+        assert "Connection to ES failed. Trying to connect to back..." in caplog.text
```
[the remark this answers is not in the document]
Comment (mariacarmina) [reply tests/test_events.py:245]: I have added but I didn't assert in the tests. I will do it now.

File: tests/test_events.py (hunk 9 above)
Comment (mariacarmina) [reply tests/test_events.py:251]: I will assert with BFACTORY_BLOCK, right?

File: setup.py [hunk 11]
```diff
@@ -23,6 +23,7 @@
     "Flask-Cors==3.0.10",
     "flask-swagger==0.2.14",
     "flask-swagger-ui==3.36.0",
+    "flask_caching==1.10.1",
```
Comment (calina-c) [inline setup.py:26]: do we need flask caching? I think we don't, it was just part of a previous iteration of the fixes.

File: pytest.ini [hunk 12]
```diff
@@ -19,4 +19,4 @@ env =
     D:NETWORK_NAME=development
     D:ADDRESS_FILE=~/.ocean/ocean-contracts/artifacts/address.json
     D:PRIVATE_KEY=0x1263dc73bef43a9da06149c7e598f52025bf4027f1d6c13896b71e81bb9233fb
-    D:SUBGRAPH_URLS={{"8996": "http://localhost:9000"}}
+    D:SUBGRAPH_URLS={{"8996": "http://localhost:9000"}}
```
Comment (calina-c) [inline pytest.ini:22]: please revert this change. There should be no change to pytest.ini.

File: tests/test_events.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (calina-c) [reply tests/test_events.py:246]: Do not forget about this one, please.

**Review by calina-c — CHANGES_REQUESTED** [review]
A few comments left.

**Review by calina-c — APPROVED** [review]
